daemon/graphdriver/copy/copy.go

```diff
@@ -9,32 +9,94 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/docker/docker/pkg/parsers/kernel"
+
+	"sync/atomic"
+
 	"github.com/docker/docker/pkg/pools"
 	"github.com/docker/docker/pkg/system"
 	rsystem "github.com/opencontainers/runc/libcontainer/system"
 	"golang.org/x/sys/unix"
 )
 
-type copyFlags int
+var copyFileRangeDisabled int32
+
+// The maximum amount of bytes Linux will splice
+const maxRwCount = 0x7ffff000
+
+// Mode indicates whether to use hardlink or copy content
+type Mode int
 
 const (
-	copyHardlink copyFlags = 1 << iota
+	// Content creates a new file, and copies the content of the file
+	Content Mode = iota
+	// Hardlink creates a new hardlink to the existing file
+	Hardlink
 )
 
-func copyRegular(srcPath, dstPath string, mode os.FileMode) error {
+func isCopyFileRangePosssible(srcDir, dstDir string) bool {
+	if copyFileRangeDisabled != 0 {
+		return false
+	}
+	// The copy_file_range syscall was introduced in kernel 4.5
+	v, err := kernel.GetKernelVersion()
+	if err != nil {
+		return false
+	}
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: 4, Major: 5, Minor: 0}) < 0 {
```

Comment:
In general one where possible should not compare kernel versions, you should just try and use the new feature and if you get `ENOSYS` (or whatever the appropriate errno is for the feature in question) apply the fallback from then on. Otherwise you miss out on backports of fixes, either via kernel.org LTS kernels or (more likely) via distros doing backports.

I see there are a couple of uses of `kernel.CompareKernelVersion` in the overlay code already but they both have an override which cases them to only produce warnings, perhaps those are for more complex fallbacks which perhaps aren't easily detectable via errno.

It actually looks like you have the majority of the fallback logic in place and are simply finding the starting point here -- it looks almost like you could just remove this check and assume available until a error code tells you otherwise (I expect there'd be some other tweaks but you are 90% of the way there I think). From something like this I don't think I'd even worry about issuing a warning...

Reply:
Do you think it's okay to use in the test to verify the behaviour of the "degradation"?

Reply:
I'm not a maintainer here, but if it were me I would simply have the tests force the copyFileRangeDisabled variable to test the degenerate case.